imports/lib/schemas/customTypes.ts

```diff
@@ -0,0 +1,43 @@
+import { Meteor } from 'meteor/meteor';
+import { z } from 'zod';
+import { Id } from './regexes';
+
+// Each of these is set to true based on the context in which the schema is
+// being evaluated. They are mutually exclusive (although they can all be false)
+export const IsInsert = new Meteor.EnvironmentVariable<boolean>();
+export const IsUpdate = new Meteor.EnvironmentVariable<boolean>();
+export const IsUpsert = new Meteor.EnvironmentVariable<boolean>();
+
+// In several of these auto-generated fields, we lie to zod about the output
+// type of the field - with type assertions, we say that they are non-nullable.
+// This will be true in the database, even though the transform functions don't
+// always return values (e.g. createdTimestamp will always be a Date, it'll just
+// get set on insertion and not on subsequent updates).
+
+export const stringId = z.string().regex(Id).optional()
```

Comment:
I'm a little fuzzy on why `stringId` is optional -- is it because it's optional when you're providing a value to e.g. `insert`, even though it's obviously a required key in every document?  What exactly do the schema types represent?

Reply:
If you look at the resulting type of `stringId`, it has an input type of `string | undefined` and an output type of `string`. The intended use of input vs. output types is connected to parsing (or validating) a schema:

```typescript
const cleaned: z.output<typeof schema> = schema.parse(raw as z.input<typeof schema>)
```

However, I'm bending those semantics a little bit, and instead using the semantic that the input type represents what must be present on a record that is passed to `insert` and the output type indicates what we expect to get back from the database. This means that `schema.parse` will not actually return the output type, but that's fine, because we are only using the return value of `schema.parse` as an intermediate anyway before sending it off to Mongo.

That means that if you're inserting a new document, you don't have to provide the `_id`, and passing the input through `parse` won't populate one, but every document you get back from the database will have one. The optional-ness of it gets stripped away by the `transform` function on the next line - the parameter `v` is of type `string | undefined`, so the casting is a lie to the type system in order to establish the input and output types as being different.